lib/integrations/awsoidc/idp_iam_config.go

```diff
@@ -154,56 +328,267 @@ func ConfigureIdPIAM(ctx context.Context, clt IdPIAMConfigureClient, req IdPIAMC
+	if err := uploadOpenIDPublicFiles(ctx, clt, req); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func ensureOIDCIdPIAM(ctx context.Context, clt IdPIAMConfigureClient, req IdPIAMConfigureRequest) error {
+	var err error
+	// For S3 bucket set ups the thumbprint is ignored, but the API still requires a parseable one.
+	// https://github.com/aws-actions/configure-aws-credentials/issues/357#issuecomment-1626357333
+	// We pass this dummy one for those scenarios.
+	thumbprint := "afafafafafafafafafafafafafafafafafafafaf"
+
+	// For set ups that use the ProxyPublicAddress, we still calculate the thumbprint.
+	if req.ProxyPublicAddress != "" {
+		thumbprint, err = ThumbprintIdP(ctx, req.ProxyPublicAddress)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	_, err = clt.CreateOpenIDConnectProvider(ctx, &iam.CreateOpenIDConnectProviderInput{
 		ThumbprintList: []string{thumbprint},
-		Url:            &req.ProxyPublicAddress,
+		Url:            &req.issuerURL,
 		ClientIDList:   []string{types.IntegrationAWSOIDCAudience},
-		Tags:           defaultResourceCreationTags(req.Cluster, req.IntegrationName).ToIAMTags(),
+		Tags:           req.ownershipTags.ToIAMTags(),
 	})
 	if err != nil {
-		if trace.IsAlreadyExists(awslib.ConvertIAMv2Error(err)) {
-			return trace.AlreadyExists("identity provider for the same URL (%s) already exists, please remove it and try again", req.ProxyPublicAddress)
+		awsErr := awslib.ConvertIAMv2Error(err)
+		if trace.IsAlreadyExists(awsErr) {
+			return nil
 		}
+
 		return trace.Wrap(err)
 	}
-	logrus.Infof("IAM OpenID Connect Provider created: url=%q arn=%q.", req.ProxyPublicAddress, aws.ToString(createOIDCResp.OpenIDConnectProviderArn))
 
-	createdIdpIAMRoleArn, err := createIdPIAMRole(ctx, clt, req)
+	return nil
+}
+
+func ensureBucketIdPIAM(ctx context.Context, clt IdPIAMConfigureClient, req IdPIAMConfigureRequest) (string, error) {
+	// According to https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html
+	// s3:GetBucketLocation is not recommended, and should be replaced by s3:HeadBucket according to AWS docs.
+	// The issue with using s3:HeadBucket is that it returns an error if the SDK client's region is not the same as the bucket.
+	// Doing a HEAD HTTP request seems to be the best option
+	resp, err := clt.HTTPHead(ctx, fmt.Sprintf("https://s3.amazonaws.com/%s", req.s3Bucket))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	// Even if the bucket is private, the "x-amz-bucket-region" Header will be there.
+	bucketRegion := resp.Header.Get("x-amz-bucket-region")
+	if bucketRegion != "" {
+		if bucketRegion == "EU" {
+			bucketRegion = "eu-west-1"
+		}
+
+		clt.SetAWSRegion(bucketRegion)
+	}
+
+	headBucketResp, err := clt.HeadBucket(ctx, &s3.HeadBucketInput{
+		Bucket:              &req.s3Bucket,
+		ExpectedBucketOwner: &req.AccountID,
+	})
+	if err == nil {
+		return aws.ToString(headBucketResp.BucketRegion), trace.AlreadyExists("")
```

Comment:
Do we need this `trace.AlreadyExists`? I would try to avoid returning error along with real relevant return value. We could use bool return parameter or pass logger into this function and let it handle logging.

Reply:
Sure :+1: 
Given this is a private method I think we would be fine.
But passing in the logger seems fine